### PR TITLE
Only update the statistics file if possible, rather than writing it anew.

### PR DESCRIPTION
**Pull request by bangerth**
Following yesterday's discussion on the forum regarding the problem @maxrudolph saw.

/rebuild

### CODE REVIEW

**Comment by bangerth** [comment]
I checked on a couple of input files that the output is only written from scratch for the first couple of time steps. After that, we only get the update operation that appends a few bytes.

**Comment by tjhei** [comment]
I am not sure I understand the logic. It looks like you never actually look at the existing statistics file but infer everything from internal data. 
What if we write a checkpoint, run two more timesteps, exit, then resume from the last checkpoint? Or did I misunderstand?

**Comment by bangerth** [comment]
On 2/25/20 8:46 PM, Timo Heister wrote:
> I am not sure I understand the logic. It looks like you never actually look at 
> the existing statistics file but infer everything from internal data.

Correct. Ingenious, eh? ;-)


> What if we write a checkpoint, run two more timesteps, exit, then resume from 
> the last checkpoint? Or did I misunderstand?

As long as you don't delete the statistics file or change it in any other way, 
everything should work -- but you'll get the two new lines twice. I see the 
problem. We should just always write the whole file upon restart.

I know how to fix this, will do tomorrow morning.

**Comment by tjhei** [comment]
> Correct. Ingenious, eh? ;-)

well... :-)

> We should just always write the whole file upon restart.

Yes. You can do this by not serializing the variables. Even better: check the file size to make sure it is identical to the last serialized size.

**Comment by tjhei** [comment]
I have two more comments:
1. Serializing potentially 100s of MB of statistics data and formatting it into a string every step is probably not ideal and fast. Not sure how to work around that if you want to be robust with checkpointing.
2. Do we need to avoid opening/closing the file every step?

**Comment by maxrudolph** [comment]
Why not just always output the entire checkpoint file after resume from
checkpoint by attaching to the appropriate signal? Then you can assume that
the statistics file is up to date and just print out the accumulated
values. You could put it within a maybe_write_statistics() function in the
same style as the other maybe_do_thisorthat() functions in core.cc and
helper_functions.cc.

**Comment by bangerth** [comment]
On 2/26/20 10:56 AM, Timo Heister wrote:
> Yes. You can do this by not serializing the variables.

That's exactly my plan. Then the variables are initialized in the 
constructor to their default values, which leads a complete write-out as 
desired.

I'll get to it this pm when back home where I have the code on my laptop.


> Even better: 
> check the file size to make sure it is identical to the last serialized 
> size.

But that requires file system operations. I think that computing the 
hash is cheaper and achieves the same result.

**Comment by bangerth** [comment]
>  1. Serializing potentially 100s of MB of statistics data and formatting
>     it into a string every step is probably not ideal and fast. Not sure
>     how to work around that if you want to be robust with checkpointing.

I suspect that it's not quite that much. @maxrudolph reports 281 bytes 
per time step, which seems reasonable. Even if you ran 100,000 time 
steps, you'd still only get to 28MB -- which is approximately the size 
of the 32MB he reports. I would expect that the overall output is much 
larger. We don't seem to see checkpointing in our profiles very much, so 
I don't think that it's an issue with run times.

It is a different issue whether we want to keep all of that data in 
memory all the time. But there, too, I don't see a lot of people running 
out of memory.

(As for the speed of formatting all of this: That's why this is running 
on a background thread.)


>  2. Do we need to avoid opening/closing the file every step?

Well, you sort of have to if you want to write data, don't you? The 
original approach tried to avoid that by just dumping everything into a 
file and then moving that file in place. But apparently that's a bad 
idea as well...

**Comment by maxrudolph** [comment]
> >   2. Do we need to avoid opening/closing the file every step?
> Well, you sort of have to if you want to write data, don't you? The
> original approach tried to avoid that by just dumping everything into a
> file and then moving that file in place. But apparently that's a bad
> idea as well...

Why not just open the statistics file once when the computation starts (or resumes) and keep it open until the program exits, allowing the writes to be buffered?

**Comment by bangerth** [comment]
Here is the updated patch, doing a complete rewrite after a restart.

File: include/aspect/simulator.h [hunk 1]
```diff
@@ -1750,6 +1750,23 @@ namespace aspect
        */
       TableHandler                        statistics;
 
+      /**
+       * The following two variables keep track which parts of the statistics
+       * object have already been written. This is because the TableHandler
+       * class has no way to keep track what it has already written, and so
+       * we can not just append the last row of the table to the output
+       * file. Rather, we keep track how many bytes we already wrote,
+       * and a has of what they contained, and if these so-many bytes have
```
Comment (tjhei) [inline include/aspect/simulator.h]:
```suggestion
       * and a hash of the contents, and if these so-many bytes have
```

File: source/simulator/helper_functions.cc [hunk 2]
```diff
@@ -289,19 +289,86 @@ namespace aspect
...
+            // case happens if the statistics_last_write_size is at the
+            // value initialized by the Simulator::Simulator()
+            // constructor, and this can happen in two situations:
+            // (i) At the end of the first time step; and (ii) at the
+            // end of the first time step after restart since the
+            // variable we query here is not serialized. It is clear that
+            // at the end of the first time step, we want to write the
+            // entire contents of the statistics object. For the second
+            // case, this is also appropriate since someone may have
+            // previously restarted from a checkpoint, run a couple of
```
Comment (tjhei) [inline source/simulator/helper_functions.cc]: This also happens without a "previous restart" if the simulation continued after writing a checkpoint (very likely).

Comment (tjhei) [inline source/simulator/helper_functions.cc]: the situation is much simpler: if you write a checkpoint and continue the simulation, more data is written into the statistics file (which will be duplicated if you resume from the checkpoint).

**Comment by bangerth** [comment]
Excellent points. I've updated the documentation in these two places!

**Comment by tjhei** [comment]
Notice that our checkpoint tests were failing in your first implementation because of duplicated lines:
https://jenkins.tjhei.info/blue/organizations/jenkins/aspect/detail/PR-3389/2/pipeline

Another win for good test coverage. Yay! :-)

**Comment by bangerth** [comment]
I didn't notice -- but yes, it's good to know!

**Comment by bangerth** [comment]
Want to merge?

**Pull request merged**